internal/modelcontroller/engine_ollama.go

```diff
@@ -0,0 +1,175 @@
+package modelcontroller
+
+import (
+	"fmt"
+	"sort"
+
+	kubeaiv1 "github.com/substratusai/kubeai/api/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func (r *ModelReconciler) oLlamaPodForModel(m *kubeaiv1.Model, c ModelConfig) *corev1.Pod {
+	lbs := labelsForModel(m)
+	ann := r.annotationsForModel(m)
+
+	if _, ok := ann[kubeaiv1.ModelPodPortAnnotation]; !ok {
+		// Set port to 8000 (vLLM) if not overwritten.
+		ann[kubeaiv1.ModelPodPortAnnotation] = "8000"
+	}
+
+	env := []corev1.EnvVar{
+		{
+			Name:  "OLLAMA_HOST",
+			Value: "0.0.0.0:8000",
+		},
+		{
+			// Ollama server typically operates in a 1:N server-to-model mode so it
+			// swaps models in and out of memory. In our case we are deploying 1:1
+			// model-to-server-pod so we want to always keep the model in memory.
+			Name: "OLLAMA_KEEP_ALIVE",
+			// Ollama treates 0 as "no keep alive" so we need to set a large value.
+			Value: "999999h",
+		},
+	}
+	var envKeys []string
+	for key := range m.Spec.Env {
+		envKeys = append(envKeys, key)
+	}
+	sort.Strings(envKeys)
+	for _, key := range envKeys {
+		env = append(env, corev1.EnvVar{
+			Name:  key,
+			Value: m.Spec.Env[key],
+		})
+	}
+
+	ollamaModelRef := c.Source.ollama.ref
+
+	featuresMap := map[kubeaiv1.ModelFeature]struct{}{}
+	for _, f := range m.Spec.Features {
+		featuresMap[f] = struct{}{}
+	}
+
+	// Pull model and copy to rename it to Model.metadata.name.
+	// See Ollama issue for rename/copy workaround: https://github.com/ollama/ollama/issues/5914
+	// NOTE: The cp command should just create a pointer to the old model, not copy data
+	// (see https://github.com/ollama/ollama/issues/5914#issuecomment-2248168474).
+	// Use `ollama run` to send a single prompt to ollama to load the model into memory
+	// before the Pod becomes Ready. (by default it will load on the first prompt request).
+	startupProbeScript := fmt.Sprintf("/bin/ollama pull %s && /bin/ollama cp %s %s",
+		ollamaModelRef, ollamaModelRef, m.Name)
+	if _, ok := featuresMap[kubeaiv1.ModelFeatureTextGeneration]; ok {
+		// NOTE: Embedding text models do not support "ollama run":
+		//
+		// ollama run nomic-embed-text hey
+		// Error: "nomic-embed-text" does not support generate
+		//
+		startupProbeScript += fmt.Sprintf(" && /bin/ollama run %s hi", m.Name)
```

Comment:
I assume the purpose of this call is to verify the model is downloaded. 
There is a simpler way to achieve the same:
```
./ollama show nomic-embed-text --modelfile
```
will return quicker with some metadata or fail with an error code when the server is not running or the model not available.

Reply:
Filed https://github.com/substratusai/kubeai/issues/277 to track to avoid scope creep